src/app/CommandHandlerDispatcher.cpp

```diff
@@ -0,0 +1,128 @@
+#include "CommandHandlerDispatcher.h"
+#include "InteractionModelEngine.h"
+#include "messaging/ExchangeContext.h"
+
+namespace chip {
+namespace app {
+using Status = Protocols::InteractionModel::Status;
+
+CHIP_ERROR CommandHandlerDispatcher::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext,
+                                                       const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err                           = CHIP_NO_ERROR;
+    bool sendStatusResponseWithInvalidAction = false;
+
+    if (mState == State::AwaitingResponse && aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        CHIP_ERROR statusError = CHIP_NO_ERROR;
+        err                    = StatusResponse::ProcessStatusResponse(std::move(aPayload), statusError);
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithInvalidAction = true);
+        statusError = err;
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithInvalidAction = true);
```

Comment:
The third of these three lines does not make sense.

Actually, neither does the second one.

This clearly needs tests at some point.

Reply:
+1

Reply:
What doesn't make sense? Pretty much I am trying to exit and respond with some sort of failure back to `CommandSender` when any of the following happens:
* We failed to Process the incoming StatusResponse
* The StatusResponse we do get is anything else other than what we expected (which is Success)

I'll agree that InvalidAction might not be the right things to send back, but I guess I don't follow what doesn't make sense about this?

Reply:
> What doesn't make sense? 

Starting with "why are we testing err twice when it can't have changed in between the tests?"

Reply:
:facepalm:  that was supposed to be `err = statusError`. Friday was not my best day, sorry